server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -479,74 +483,97 @@ public void cloneShardSnapshot(
+                    existingShardGen = shardGeneration;
+                }
+                SnapshotFiles existingTargetFiles = null;
+                SnapshotFiles sourceFiles = null;
+                for (SnapshotFiles existingSnapshot : existingSnapshots) {
+                    final String snapshotName = existingSnapshot.snapshot();
+                    if (snapshotName.equals(target.getName())) {
+                        existingTargetFiles = existingSnapshot;
+                    } else if (snapshotName.equals(source.getName())) {
+                        sourceFiles = existingSnapshot;
+                    }
+                    if (sourceFiles != null && existingTargetFiles != null) {
+                        break;
+                    }
+                }
+                if (sourceFiles == null) {
+                    throw new RepositoryException(
+                        metadata.name(),
+                        "Can't create clone of ["
+                            + shardId
+                            + "] for snapshot ["
+                            + target
+                            + "]. The source snapshot ["
+                            + source
+                            + "] was not found in the shard metadata."
+                    );
+                }
+                if (existingTargetFiles != null) {
+                    if (existingTargetFiles.isSame(sourceFiles)) {
+                        return existingShardGen;
+                    }
+                    throw new RepositoryException(
+                        metadata.name(),
+                        "Can't create clone of ["
+                            + shardId
+                            + "] for snapshot ["
+                            + target
+                            + "]. A snapshot by that name already exists for this shard."
+                    );
+                }
+                final BlobStoreIndexShardSnapshot sourceMeta = loadShardSnapshot(shardContainer, source);
+                logger.trace("[{}] [{}] writing shard snapshot file for clone", shardId, target);
+                INDEX_SHARD_SNAPSHOT_FORMAT.write(
+                    sourceMeta.asClone(target.getName(), startTime, threadPool.absoluteTimeInMillis() - startTime),
+                    shardContainer,
+                    target.getUUID(),
+                    compress
+                );
+                INDEX_SHARD_SNAPSHOTS_FORMAT.write(
+                    existingSnapshots.withClone(source.getName(), target.getName()),
+                    shardContainer,
+                    newGen,
+                    compress
                 );
-                existingShardGen = String.valueOf(tuple.v2());
-                newGen = String.valueOf(tuple.v2() + 1);
-                existingSnapshots = tuple.v1();
             } else {
                 newGen = UUIDs.randomBase64UUID();
```

Comment:
why are we creating a `newGen` for the shallow snapshot copy ? This should ideally return the `existingShardGen` as shallow copy doesn't change the shard gen.